include/llvm-dialects/Dialect/OpMap.h

```diff
@@ -0,0 +1,356 @@
+        if (dialectOpKV == desc)
+          return dialectOpKV.Value;
+      }
+    }
+
+    // Nothing found. Insert a new entry in the correct map.
+    return insert_or_get(desc, {});
+  }
+
+  // -------------------------------------------------------------
+  // Convenience operator[] definitions.
+  // -------------------------------------------------------------
+  ValueT &operator[](const IntrinsicInst &inst) {
+    return m_intrinsics[inst.getIntrinsicID()];
+  }
+
+  ValueT &operator[](const OpDescription &desc) { return get(desc); }
+
+  // -------------------------------------------------------------
+  // lookup tries to find whether a given function or instruction
+  // can be mapped to any of the entries in the internal map
+  // containers.
+  // It returns either a default-constructed object if the key
+  // was not found or a copy of the contained value.
+  // -------------------------------------------------------------
+
+  template <typename OpT> ValueT lookup(const Function &func) const {
+    return lookup(OpDescription::get<OpT>(), func);
+  }
+
+  ValueT lookup(const OpDescription &desc, const Function &func) const {
+    if (desc.isCoreOp())
+      return {};
+
+    if (desc.isIntrinsic() && func.isIntrinsic()) {
+      if (m_intrinsics.contains(func.getIntrinsicID()) &&
+          desc.matchDeclaration(func))
+        return m_intrinsics.lookup(func.getIntrinsicID());
+    }
+
+    for (const DialectOpKV &dialectOpKV : m_dialectOps) {
+      if (dialectOpKV == desc && desc.matchDeclaration(func))
+        return dialectOpKV.Value;
+    }
+
+    return {};
+  }
+
+  template <typename OpT> ValueT lookup(const Instruction &inst) const {
+    return lookup(OpDescription::get<OpT>(), inst);
+  }
+
+  ValueT lookup(const OpDescription &desc, const Instruction &inst) const {
+    if (isa<IntrinsicInst>(inst) || isa<CallInst>(inst))
+      return lookup(*cast<CallInst>(inst).getCalledFunction());
+
+    if (!desc.isCoreOp())
+      return {};
+
+    if (m_coreOpcodes.contains(inst.getOpcode()) && desc.matchInstruction(inst))
+      return m_coreOpcodes.lookup(inst.getOpcode());
+
+    return {};
+  }
+
+  // Try to lookup a function which is either the callee of an intrinsic call
+  // or a dialect operation.
+  ValueT lookup(const Function &func) const {
+    if (func.isIntrinsic())
+      return m_intrinsics.lookup(func.getIntrinsicID());
+
+    for (const DialectOpKV &dialectOpKV : m_dialectOps) {
+      const DialectOpKey &key = dialectOpKV.Key;
+      if (isMatchingFunc(key.first, key.second, func.getName()))
+        return dialectOpKV.Value;
+    }
+
+    return {};
+  }
+
+  // Try to lookup an instruction which is either an intrinsic instruction,
+  // a dialect operation or a core instruction.
+  ValueT lookup(const Instruction &inst) const {
+    if (isa<IntrinsicInst>(inst) || isa<CallInst>(inst))
+      return lookup(*cast<CallInst>(inst).getCalledFunction());
```

Comment:
This can just be `isa<CallInst>`, since `IntrinsicInst` is derived from `CallInst`.

I wonder whether this case should fallback to looking up `OpDescription::get<CallInst>()`. I tend to think that it should, though in any case it feels likely that users of this functionality will always need some amount of special-casing around `CallInst`.

Note that this also needs to handle the case that `getCalledFunction` returns nullptr.